specfile/sources.py

```diff
@@ -0,0 +1,466 @@
+# Copyright Contributors to the Packit project.
+# SPDX-License-Identifier: MIT
+
+import re
+import urllib.parse
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Iterable, List, Optional, Tuple, Union, overload
+
+from specfile.exceptions import SpecfileException
+from specfile.rpm import Macros
+from specfile.sourcelist import Sourcelist, SourcelistEntry
+from specfile.tags import Comments, Tag, Tags
+
+
+class Source(ABC):
+    """Class that represents a source."""
+
+    @property  # type: ignore
+    @abstractmethod
+    def index(self) -> int:
+        """Numeric index of the source."""
+        pass
+
+    @property  # type: ignore
+    @abstractmethod
+    def location(self) -> str:
+        """Literal location of the source as stored in the spec file."""
+        pass
+
+    @location.setter  # type: ignore
+    @abstractmethod
+    def location(self, value: str) -> None:
+        pass
+
+    @property  # type: ignore
+    @abstractmethod
+    def expanded_location(self) -> str:
+        """Location of the source after expanding macros."""
+        pass
+
+    @property  # type: ignore
+    @abstractmethod
+    def filename(self) -> str:
+        """Literal filename of the source."""
+        pass
+
+    @property  # type: ignore
+    @abstractmethod
+    def expanded_filename(self) -> str:
+        """Filename of the source after expanding macros."""
+        pass
+
+    @property  # type: ignore
+    @abstractmethod
+    def comments(self) -> Comments:
+        """List of comments associated with the source."""
+        pass
+
+
+class TagSource(Source):
+    """Class that represents a source backed by a spec file tag."""
+
+    def __init__(self, tag: Tag) -> None:
+        """
+        Constructs a `TagSource` object.
+
+        Args:
+            tag: Tag that this source represents.
+
+        Returns:
+            Constructed instance of `TagSource` class.
+        """
+        self._tag = tag
+
+    def __repr__(self) -> str:
+        tag = repr(self._tag)
+        return f"TagSource({tag})"
+
+    def _get_index(self) -> Optional[str]:
+        """
+        Extracts numeric index from tag name.
+
+        Returns:
+            Extracted index or None if there isn't one.
+        """
+        tokens = re.split(r"(\d+)", self._tag.name, maxsplit=1)
+        if len(tokens) > 1:
+            return tokens[1]
+        return None
+
+    @property
+    def index(self) -> int:
+        """Numeric index of the source."""
+        return int(self._get_index() or 0)
+
+    @property
+    def index_digits(self) -> int:
+        """Number of digits the index is formed by."""
+        return len(self._get_index() or "")
+
+    @property
+    def location(self) -> str:
+        """Literal location of the source as stored in the spec file."""
+        return self._tag.value
+
+    @location.setter
+    def location(self, value: str) -> None:
+        self._tag.value = value
```

Comment:
It does, but that's a problem in `Tag`. To be honest, I was thinking about dropping the expanded value and validity completely, as it can't be guaranteed to be correct. We could have a dynamic `Tag.expanded_value`, implemented using `Macros.expand()`. What do you think about it?